### ISSUE TITLE
My servers used for multi-node training do not have ssh. How can I launch multi-node training using the torchrun command?

### ISSUE THREAD

**Issue by dingning97** [EleutherAI/gpt-neox#1203]
My servers used for multi-node training do not have ssh. How can I launch multi-node training using the torchrun command?

**Comment by WKX933** [comment]
I also encountered the same problem. Have you found a solution?

**Comment by Quentin-Anthony** [comment]
This can be resolved by adding `torchrun` as a possible deepspeed multinode runner. We're targeting adding this within the next week or two.

**Comment by ameyagodbole** [comment]
I am in a similar situation. I am working on a cluster where the code needs to be containerized, and SSH inter-node communication is not possible.

I realized that `gpt-neox/deepy.py` is trying to launch the multi-node jobs (via PDSH/SLURM). But in my setup, this is not possible since I need to launch the containers with `srun`, and the containers themselves don't have a way to communicate with the slurm controller or SSH to other nodes.

I have written this un-wrapper to skip the deepspeed launcher while doing the argument processing that `gpt-neox/deepy.py` does. Can someone please validate if this is correct? I am hesitant to make this an official pull request since it is specialized for SLURM and does not handle the other deepspeed launch mechanisms (PDSH, OpenMPI, MVAPICH, MPICH, IMPI).

https://github.com/ameyagodbole/hubble-gpt-neox/blob/hubble-skip-deepspeed-update/deepy_simple.py

**Comment by zwzz2019** [comment]
Here is our solution: we have successfully trained the model in a multi-node setup with the following modifications.
Below, I will share the code and instructions — please adjust them according to your own environment.

1. **Modify the launch script**:  
   Update the bash script used to start training:  
   [node8.sh](https://github.com/zwzz2019/neox/blob/main/node8.sh) 

2. **Update the YAML configuration file**:  
   Edit the corresponding YAML config file:  
   [node8.yml](https://github.com/zwzz2019/neox/blob/main/node8.yml)   
   Pay particular attention to the settings for `no-ssh` and `rank`.

3. **Patch the DeepSpeed runner.py**:  
   Modify the `runner.py` file located in the DeepSpeed launcher directory (`/deepspeed/launcher/`):  
   [runner.py](https://github.com/zwzz2019/neox/blob/main/runner.py)   
   Add preprocessing logic to handle the `--node_rank` argument correctly.

Please feel free to adapt these steps based on your cluster configuration.

**Comment by ameyagodbole** [comment]
Thanks for the minimal example!

So to clarify: the key is to use PDSH launcher, but set the node_rank to always be 0 for deepspeed launcher?

FYI, we have been running with a different workaround, which is to use torchrun to launch experiments.

**Comment by zwzz2019** [comment]
No, you need to specify the node rank parameter when starting bash. My example is applicable for cases where pdsh cannot be used. 

If torchrun is available, that's even better since it supports many features.
